File: api/src/main/java/com/ruby/rubia_server/core/service/DepartmentService.java
package com.ruby.rubia_server.core.service;

import com.ruby.rubia_server.core.dto.CreateDepartmentDTO;
import com.ruby.rubia_server.core.dto.DepartmentDTO;
import com.ruby.rubia_server.core.dto.UpdateDepartmentDTO;
import com.ruby.rubia_server.core.entity.Company;
import com.ruby.rubia_server.core.entity.Department;
import com.ruby.rubia_server.core.repository.CompanyRepository;
import com.ruby.rubia_server.core.repository.DepartmentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class DepartmentService {
    
    private final DepartmentRepository departmentRepository;
    private final CompanyRepository companyRepository;
    
    public DepartmentDTO create(CreateDepartmentDTO createDTO, UUID companyId) {
        log.info("Creating department with name: {} for company: {}", createDTO.getName(), companyId);
        
        if (departmentRepository.existsByNameAndCompanyId(createDTO.getName(), companyId)) {
            throw new IllegalArgumentException("Departamento com nome '" + createDTO.getName() + "' já existe");
        }
        
        Company company = companyRepository.findById(companyId)
                .orElseThrow(() -> new IllegalArgumentException("Empresa não encontrada"));
        
        Department department = Department.builder()
                .name(createDTO.getName())
                .description(createDTO.getDescription())
                .company(company)
                .autoAssign(createDTO.getAutoAssign())
                .build();
        
        Department saved = departmentRepository.save(department);
        log.info("Department created successfully with id: {}", saved.getId());
        
        return toDTO(saved);
    }
    
    @Transactional(readOnly = true)
    public DepartmentDTO findById(UUID id, UUID companyId) {
        log.debug("Finding department by id: {} for company: {}", id, companyId);
        
        Department department = departmentRepository.findByIdAndCompanyId(id, companyId)
                .orElseThrow(() -> new IllegalArgumentException("Departamento não encontrado"));
        
        return toDTO(department);
    }
    
    @Transactional(readOnly = true)
    public List<DepartmentDTO> findAll(UUID companyId) {
        log.debug("Finding all departments for company: {}", companyId);
        
        return departmentRepository.findAllByCompanyIdOrderedByName(companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    @Transactional(readOnly = true)
    public List<DepartmentDTO> findByAutoAssign(UUID companyId) {
        log.debug("Finding departments with auto assign enabled for company: {}", companyId);
        
        return departmentRepository.findByAutoAssignTrueAndCompanyId(companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    public DepartmentDTO update(UUID id, UpdateDepartmentDTO updateDTO, UUID companyId) {
        log.info("Updating department with id: {} for company: {}", id, companyId);
        
        Department department = departmentRepository.findByIdAndCompanyId(id, companyId)
                .orElseThrow(() -> new IllegalArgumentException("Departamento não encontrado"));
        
        if (updateDTO.getName() != null) {
            if (!updateDTO.getName().equals(department.getName()) && 
                departmentRepository.existsByNameAndCompanyId(updateDTO.getName(), companyId)) {
                throw new IllegalArgumentException("Departamento com nome '" + updateDTO.getName() + "' já existe");
            }
            department.setName(updateDTO.getName());
        }
        
        if (updateDTO.getDescription() != null) {
            department.setDescription(updateDTO.getDescription());
        }
        
        if (updateDTO.getAutoAssign() != null) {
            department.setAutoAssign(updateDTO.getAutoAssign());
        }
        
        Department updated = departmentRepository.save(department);
        log.info("Department updated successfully");
        
        return toDTO(updated);
    }
    
    public void delete(UUID id, UUID companyId) {
        log.info("Deleting department with id: {} for company: {}", id, companyId);
        
        if (!departmentRepository.existsByIdAndCompanyId(id, companyId)) {
            throw new IllegalArgumentException("Departamento não encontrado");
        }
        
        departmentRepository.deleteByIdAndCompanyId(id, companyId);
        log.info("Department deleted successfully");
    }
    
    private DepartmentDTO toDTO(Department department) {
        return DepartmentDTO.builder()
                .id(department.getId())
                .name(department.getName())
                .description(department.getDescription())
                .autoAssign(department.getAutoAssign())
                .createdAt(department.getCreatedAt())
                .updatedAt(department.getUpdatedAt())
                .build();
    }
}

File: api-dotnet/Rubia.Server/Services/DepartmentService.cs
using Microsoft.EntityFrameworkCore;
using Rubia.Server.Data;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Services.Interfaces;

namespace Rubia.Server.Services;

public class DepartmentService : IDepartmentService
{
    private readonly RubiaDbContext _context;
    private readonly ILogger<DepartmentService> _logger;

    public DepartmentService(RubiaDbContext context, ILogger<DepartmentService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<DepartmentDto> CreateAsync(CreateDepartmentDto createDto, Guid companyId)
    {
        _logger.LogInformation("Creating department with name: {Name} for company: {CompanyId}", 
            createDto.Name, companyId);

        if (await _context.Departments.AnyAsync(d => d.Name == createDto.Name && d.CompanyId == companyId))
        {
            throw new ArgumentException($"Departamento com nome '{createDto.Name}' já existe");
        }

        var company = await _context.Companies.FirstOrDefaultAsync(c => c.Id == companyId);
        if (company == null)
        {
            throw new ArgumentException("Empresa não encontrada");
        }

        var department = new Department
        {
            Name = createDto.Name,
            Description = createDto.Description,
            CompanyId = companyId,
            AutoAssign = createDto.AutoAssign
        };

        _context.Departments.Add(department);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Department created successfully with id: {Id}", department.Id);
        return ToDto(department);
    }

    public async Task<DepartmentDto> FindByIdAsync(Guid id, Guid companyId)
    {
        _logger.LogDebug("Finding department by id: {Id} for company: {CompanyId}", id, companyId);

        var department = await _context.Departments
            .FirstOrDefaultAsync(d => d.Id == id && d.CompanyId == companyId);

        if (department == null)
        {
            throw new ArgumentException("Departamento não encontrado");
        }

        return ToDto(department);
    }

    public async Task<List<DepartmentDto>> FindAllAsync(Guid companyId)
    {
        _logger.LogDebug("Finding all departments for company: {CompanyId}", companyId);

        var departments = await _context.Departments
            .Where(d => d.CompanyId == companyId)
            .OrderBy(d => d.Name)
            .ToListAsync();

        return departments.Select(ToDto).ToList();
    }

    public async Task<List<DepartmentDto>> FindByAutoAssignAsync(Guid companyId)
    {
        _logger.LogDebug("Finding departments with auto assign enabled for company: {CompanyId}", companyId);

        var departments = await _context.Departments
            .Where(d => d.AutoAssign && d.CompanyId == companyId)
            .OrderBy(d => d.Name)
            .ToListAsync();

        return departments.Select(ToDto).ToList();
    }

    public async Task<DepartmentDto> UpdateAsync(Guid id, UpdateDepartmentDto updateDto, Guid companyId)
    {
        _logger.LogInformation("Updating department with id: {Id} for company: {CompanyId}", id, companyId);

        var department = await _context.Departments
            .FirstOrDefaultAsync(d => d.Id == id && d.CompanyId == companyId);

        if (department == null)
        {
            throw new ArgumentException("Departamento não encontrado");
        }

        if (updateDto.Name != null)
        {
            if (updateDto.Name != department.Name && 
                await _context.Departments.AnyAsync(d => d.Name == updateDto.Name && d.CompanyId == companyId))
            {
                throw new ArgumentException($"Departamento com nome '{updateDto.Name}' já existe");
            }
            department.Name = updateDto.Name;
        }

        if (updateDto.Description != null)
        {
            department.Description = updateDto.Description;
        }

        if (updateDto.AutoAssign.HasValue)
        {
            department.AutoAssign = updateDto.AutoAssign.Value;
        }

        await _context.SaveChangesAsync();
        _logger.LogInformation("Department updated successfully");

        return ToDto(department);
    }

    public async Task DeleteAsync(Guid id, Guid companyId)
    {
        _logger.LogInformation("Deleting department with id: {Id} for company: {CompanyId}", id, companyId);

        var department = await _context.Departments
            .FirstOrDefaultAsync(d => d.Id == id && d.CompanyId == companyId);

        if (department == null)
        {
            throw new ArgumentException("Departamento não encontrado");
        }

        _context.Departments.Remove(department);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Department deleted successfully");
    }

    private static DepartmentDto ToDto(Department department)
    {
        return new DepartmentDto
        {
            Id = department.Id,
            CompanyId = department.CompanyId,
            Name = department.Name,
            Description = department.Description,
            AutoAssign = department.AutoAssign,
            CreatedAt = department.CreatedAt,
            UpdatedAt = department.UpdatedAt
        };
    }
}